442 lfric openacc psy

rupertford opened the pull request.

rupertford commented:
This is now ready for 1st review (requested @arporter).

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -32,7 +32,8 @@
 # -----------------------------------------------------------------------------
 # Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
-# Modified I. Kavcic Met Office, C.M. Maynard, Met Office/University of Reading.
+# Modified I. Kavcic Met Office
```
Comma after "Kavcic" please.

rupertford replied:
Changed (and added space between middle names for Duke Maynard).

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6616,40 +6616,123 @@ def test_intergrid_err(dist_mem):
+    _ = acc_enter_trans.apply(sched)
+    code = str(psy.gen)
+    assert (
+        "      !$acc enter data\n"
```
Why doesn't this `enter data` specify what data to move to the device? The kernel must use at least one field and that should be listed unless I'm missing something? Or is it that there are no OpenACC parallel or kernels directives? In which case I think the code-generation step should raise an exception as it makes no sense to have one without the other.

rupertford replied:
It is the latter. Enter Data looks at the ACCParallel directives within its region to determine data to move. I've added a check at code generation time.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -6616,40 +6616,123 @@ def test_intergrid_err(dist_mem):
+    '''
+    Test that an ACCParallelTrans transformation can add an OpenACC
+    Parallel directive to the PSy layer in the dynamo0.3 API. An
+    EnterData directive is also required otherwise the transformation
```
"otherwise an exception is raised at code-generation time." I think.

rupertford replied:
Yes. Clarified.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1581,44 +1581,6 @@ def test_acc_parallel_invalid_node():
-    assert "LOGICAL, save :: first_time=.True." in code
-
-    pcopy = (
-        "      IF (first_time) THEN\n"
```
I know we no longer have the IF(first_time)THEN bit but it seems a bit strong to delete the whole test? Is the functionality being tested now in psyGen?

rupertford replied:
You're right, I was over zealous. I've reinstated the test with the appropriate part.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1655,111 +1617,6 @@ def test_acc_data_grid_copyin(tmpdir):
     assert GOcean1p0Build(tmpdir).code_compiles(psy)
 
 
-def test_acc_rscalar_update(tmpdir):
```
Again, we need to check that we can safely do away with this functionality. Maybe the fact that we now always do an `enter data` (i.e. there's no `if(first_time)...`) *and* include scalars in it covers us - I think it should.

rupertford replied:
As mentioned before, I'll test this.

rupertford replied:
The code has been tested using example gocean/eg2 on glados and correct results obtained with scalars (both constant values and changing values). 55.0 is the magic number (1+2+....+10). Thanks to @arporter for helping with this. Running this example did demonstrate a separate problem which is captured in issue #491.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2107,18 +2103,227 @@ def test_acc_dag_names():
+# (4/4) Method gen_code
+def test_accenterdatadirective_gencode_4():
+    '''Test that an OpenACC Enter Data directive, when added to a schedule
+    with a multiple loops and multiple OpenACC parallel directives,
```
s/with a/with/

rupertford replied:
Well spotted. Cut n paste strikes again.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
 
   > python find_kernel.py
 
+Example 14
```
Done. I just added OpenACC to be consistent with the GOcean examples.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
+expected. However it is never-the-less useful as a starting
+point. Three scripts are provided.
+
+The first script shows how to add OpenACC Kernels directives to the
```
Good point. Done.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
+The second script shows how to add OpenACC Loop, Parallel and Entry
+Data directives to the PSy-layer. Again this example only works with
+distributed memory switched off as the OpenACC Parallel transformation
+does not support halo exchanges within an OpenACC Parallel directive.
```
We spotted. There are a few more so I've done those too.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
+off as the OpenACC Kernels transformation does not yet support halo
+exchanges within an OpenACC Kernels directive.
+
+The second script shows how to add OpenACC Loop, Parallel and Entry
```
Oops.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
+distributed memory being switched on by placing an OpenACC Parallel
+directive around each OpenACC Loop directive, rather than having one
+for the whole invoke. This approach avoids having halo exchanges
+within am OpenACC Parallel directive.
```
Done.

rupertford commented on doc/user_guide/examples.rst:
```diff
@@ -190,6 +190,38 @@ to the specified Kernel:
+
+The generated code has a number of problems including 1) it does not
+modify the kernels to include the OpenACC Routine directive, 2) a
+loops upper bounds is computed via a derived type (this should be
```
S's were never my strong point.

rupertford commented on examples/dynamo/README:
```diff
@@ -264,6 +264,33 @@ code
+
+1) Adding OpenACC kernels directives. -nodm is used as an exception is
+raised if haloexchanges are found within an OpenACC kernels region.
+::
```
I'm not saying it is right but I just copied the other 13 examples which use the same (Sphinx) formatting. Are you suggesting we change README to README.md. If so, what is the benefit of this - does the browser display it in markdown form when it is viewed?

arporter replied:
Yes and yes!

rupertford replied:
OK, I've done this.

rupertford commented on examples/dynamo/README:
```diff
@@ -264,6 +264,33 @@ code
+evaluation.
+
+1) Adding OpenACC kernels directives. -nodm is used as an exception is
+raised if haloexchanges are found within an OpenACC kernels region.
```
Done.

arporter replied:
Almost ;-) s/Exchanges/Exchange/ and s/withing/within/

rupertford replied:
Oops, butter fingers.

rupertford commented on examples/dynamo/eg14/acc_kernels.py:
```diff
@@ -0,0 +1,59 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. Ford, STFC Daresbury Laboratory
```
I must have cut and paste'd from somewhere. Updated.

rupertford commented on examples/dynamo/eg14/acc_parallel_dm.py:
```diff
@@ -0,0 +1,66 @@
+
+'''File containing a PSyclone transformation script for the Dynamo0p3
+API to apply OpenACC Loop, Parallel and Enter Data directives
+generically. This can be applied via the -s option in the psyclone
```
Done. I wasn't sure whether to capitalise to Halo Exchanges and opted for your suggested lower case.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8245,6 +8245,121 @@ def raw_arg_list(self):
+            def fs_compulsory_field(self, function_space):
+                '''
+                Add compulsory arguments associated with this function space to
+                the list.  argument. OpenACC requires the full
```
Oops.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1999,50 +1999,18 @@ def gen_code(self, parent):
-            for var in pdir.scalars:
-                if var not in var_list:
-                    var_list.append(var)
-        if var_list:
```
No I'm not confident, I'm just going on what we discussed. I'll test it when I'm back in the office.

rupertford replied:
Now tested, please see later comment.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -1633,12 +1614,11 @@ class ACCParallelTrans(ParallelRegionTrans):
-                        psyGen.ACCLoopDirective, psyGen.Assignment,
-                        psyGen.Reference, psyGen.Literal,
-                        psyGen.BinaryOperation)
+    from psyclone import gocean1p0, nemo, psyGen, dynamo0p3
```
Done.

rupertford commented:
@arporter. I've addressed all comments except for testing that the modifications don't break the gocean benchmark and asking a question about a README. I think it is ready for next review but am happy if you want to wait until I check that the benchmark still works.

arporter commented on examples/dynamo/eg14/acc_parallel_dm.py:
```diff
@@ -31,12 +31,12 @@
-generically. This can be applied via the -s option in the psyclone
-script.
+generically in the presence of halo exchanges. This can be applied via
+the -s option in the psyclone script.
```
s/in the/to the/

rupertford replied:
Good spot.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2000,14 +2001,15 @@ def dag_name(self):
         '''
-        from psyclone.f2pygen import DeclGen, DirectiveGen, CommentGen, \
-            IfThenGen, AssignGen, CallGen, UseGen
+        from psyclone.f2pygen import DeclGen, DirectiveGen, CommentGen
```
pylint complains that `DirectiveGen` has already been imported.

rupertford replied:
Removed this and all other DIrectiveGen imports in this file that are not required.

arporter commented on src/psyclone/tests/nemo/transformations/openacc/kernels_directive_test.py:
```diff
@@ -56,20 +56,6 @@
                  "end program do_loop\n")
 
 
```
pylint complains that `InternalError` is imported but not used.

rupertford replied:
Removed.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2867,8 +2832,10 @@ def _validate(self, sched):
                          (subclass of) :py:class:`psyclone.psyGen.Schedule`.
         '''
         from psyclone.psyGen import Directive, \
```
pylint complains that `ACCKernelsDirective` and `ACCParallelDirective` are not used.

rupertford replied:
Done.

rupertford commented:
Ready for next review from @arporter

arporter commented:
All requested changes have been made. Examples README has been converted to MarkDown and renders nicely now. The new code has been tested with the OpenACC example and works fine (modulo a separate problem with directives in the Algorithm layer).
Issue 442 is not referenced in any TODOs.
All tests pass (including with compilation).

arporter commented:
Documentation has been updated and builds fine.
